libraries/ms-common/test/test_adc.c

```diff
@@ -0,0 +1,148 @@
+                    adc_register_callback(NUM_ADC_CHANNEL, prv_callback, NULL));
+
+  TEST_ASSERT_EQUAL(STATUS_CODE_EMPTY, adc_register_callback(ADC_CHANNEL_0, prv_callback, NULL));
+  TEST_ASSERT_EQUAL(STATUS_CODE_EMPTY, adc_register_callback(ADC_CHANNEL_1, prv_callback, NULL));
+  TEST_ASSERT_EQUAL(STATUS_CODE_EMPTY, adc_register_callback(ADC_CHANNEL_2, prv_callback, NULL));
+
+  adc_set_channel(ADC_CHANNEL_0, 1);
+  adc_set_channel(ADC_CHANNEL_1, 1);
+  adc_set_channel(ADC_CHANNEL_2, 1);
+
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, adc_register_callback(ADC_CHANNEL_0, prv_callback, NULL));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, adc_register_callback(ADC_CHANNEL_1, prv_callback, NULL));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, adc_register_callback(ADC_CHANNEL_2, prv_callback, NULL));
+}
+
+void test_single() {
+  uint16_t reading;
+
+  // Initialize the ADC to single mode and configure the channels
+  adc_init(ADC_MODE_SINGLE);
+
+  adc_set_channel(ADC_CHANNEL_0, 1);
+  adc_set_channel(ADC_CHANNEL_1, 1);
+  adc_set_channel(ADC_CHANNEL_2, 1);
+
+  adc_register_callback(ADC_CHANNEL_0, prv_callback, NULL);
+  adc_register_callback(ADC_CHANNEL_1, prv_callback, NULL);
+  adc_register_callback(ADC_CHANNEL_2, prv_callback, NULL);
+
+  // Callbacks must not run in single mode unless a read occurs
+  TEST_ASSERT_FALSE(s_callback_ran);
+  TEST_ASSERT_EQUAL(0, s_callback_runs);
+
+  // Ensure that the conversions happen once adc_read_value is called
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, adc_read_raw(ADC_CHANNEL_0, &reading));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, adc_read_raw(ADC_CHANNEL_1, &reading));
+  TEST_ASSERT_EQUAL(STATUS_CODE_OK, adc_read_raw(ADC_CHANNEL_2, &reading));
+
+  TEST_ASSERT_EQUAL(STATUS_CODE_INVALID_ARGS, adc_read_raw(NUM_ADC_CHANNEL, &reading));
+  TEST_ASSERT_EQUAL(STATUS_CODE_EMPTY, adc_read_raw(ADC_CHANNEL_3, &reading));
+
+  while (!s_callback_ran) {}
+
+  TEST_ASSERT_TRUE(s_callback_ran);
+  TEST_ASSERT_TRUE(s_callback_runs > 0);
+  TEST_ASSERT_TRUE(0 <= reading && reading < 4095);
+}
+
+void test_continuous() {
+  s_callback_runs = 0;
+  s_callback_ran = false;
+
+  // Initialize ADC and check that adc_init() can properly reset the ADC
+  adc_init(ADC_MODE_CONTINUOUS);
+
+  adc_set_channel(ADC_CHANNEL_0, 1);
+  adc_set_channel(ADC_CHANNEL_1, 1);
+  adc_set_channel(ADC_CHANNEL_2, 1);
+
+  adc_register_callback(ADC_CHANNEL_0, prv_callback, NULL);
+  adc_register_callback(ADC_CHANNEL_1, prv_callback, NULL);
+  adc_register_callback(ADC_CHANNEL_2, prv_callback, NULL);
+
+  // Run a busy loop until a callback is triggered
+  while (!s_callback_runs) { }
+
+  TEST_ASSERT_TRUE(s_callback_ran);
+  TEST_ASSERT_TRUE(s_callback_runs > 0);
+}
+
+void test_read_single() {
+  // Check that both the raw readings and converted readings are within the expected range
+  adc_init(ADC_MODE_SINGLE);
+
+  adc_set_channel(ADC_CHANNEL_0, 1);
+  adc_register_callback(ADC_CHANNEL_0, prv_callback, NULL);
+
+  prv_adc_read(ADC_CHANNEL_0);
```

Comment:
Might want to rename this so that its function is more obvious. I was a little confused about where you were checking for values within the expected range until I scrolled up.